hooks/pre-command

```diff
@@ -29,6 +31,29 @@ check_set() {
   fi
 }
 
+maybe_download_sc_archive() {
+  local sauce_connect_version="$1"
+  local os="$2"
+
+  local sc_archive
+  case "${os}" in
+    linux*)
+      sc_archive="sc-${sauce_connect_version}-${os}.tar.gz"
+    ;;
+    darwin*)
+      sc_archive="sc-${sauce_connect_version}-${os}.zip"
+    ;;
+    *) error "unknown OS: ${os}" ;;
+  esac
+
+  local sc_archive_path="${TMP_DIR}/${PLUGIN_NAME}/${sc_archive}"
+  if [[ ! -e "${sc_archive_path}" ]]; then
+    curl -fsSL --create-dirs --retry 5 -o "${sc_archive_path}" "https://saucelabs.com/downloads/${sc_archive}"
+  fi
```

Comment:
It's dangerous to download directly to the path where we're going to cache it: if the download is interrupted we are left with a corrupted file, and instead of resuming the download with curl we assume it's fully downloaded and proceed to extract from it.

Instead download it to a temporary file path created with `mktemp` in the same directory, then `mv` it to the right path. File renames on the same posix filesystem are atomic, so the file won't get corrupted, and since you're in the same directory you're guaranteed to be on the same filesystem.

Make sure you set up a cleanup trap for the temporary file so in case the download is interrupted we delete it instead of leaving it behind (this might not actually happen in practice depending on how brutally the script gets interrupted, but it's still good practice).

Reply:
> It's dangerous to download directly to the path ... if the download is interrupted we are left with a corrupted file

TIL! Fixed in 9cde255863f0f9e740a533e7540a60996017f26b.

Is this a curl thing? That is, will this happen if the connection is reset mid-download for example? Or would this happen if the process was killed?

Reply:
I think it will happen in both cases. Curl supports resuming downloads, so that's not a bad thing, you just have to handle it.